src/lib/utils/bitvector.h

```diff
@@ -0,0 +1,1327 @@
+                     result_block |= block << bits_in_carry;
+                  }
+               }
+
+               m_read_bitpos += std::min(block_bits, bits_remaining);
+               return result_block;
+            }
+
+            /**
+             * Stores the next block of bits into the underlying bitvector.
+             * No bounds checks are performed. Storing bit blocks that are not
+             * aligned at a byte-boundary in the underlying bitvector is
+             * currently not implemented.
+             */
+            template <std::unsigned_integral BlockT>
+               requires(!is_const())
+            void store_next(BlockT block) {
+               constexpr size_type block_size = sizeof(BlockT);
+               constexpr size_type block_bits = block_size * 8;
+
+               if constexpr(is_byte_aligned()) {
+                  auto sink = m_source.as_byte_span().subspan(write_bytepos()).template first<block_size>();
+                  store_le(sink, block);
+                  m_write_bitpos += std::min(block_bits, bits_to_write());
+               } else {
+                  throw Not_Implemented("Storing out-of-alignment blocks is NYI");
+               }
+            }
+
+            template <std::unsigned_integral BlockT>
+               requires(is_byte_aligned() && !is_const())
+            std::span<BlockT> span(size_type blocks) const {
+               BOTAN_DEBUG_ASSERT(blocks == 0 || is_memory_aligned_to<BlockT>());
+               BOTAN_DEBUG_ASSERT(read_bytepos() % sizeof(BlockT) == 0);
+               // Intermittently casting to void* to avoid a compiler warning
+               void* ptr = reinterpret_cast<void*>(m_source.as_byte_span().data() + read_bytepos());
+               return {reinterpret_cast<BlockT*>(ptr), blocks};
+            }
+
+            template <std::unsigned_integral BlockT>
+               requires(is_byte_aligned() && is_const())
+            std::span<const BlockT> span(size_type blocks) const {
+               BOTAN_DEBUG_ASSERT(blocks == 0 || is_memory_aligned_to<BlockT>());
+               BOTAN_DEBUG_ASSERT(read_bytepos() % sizeof(BlockT) == 0);
+               // Intermittently casting to void* to avoid a compiler warning
+               const void* ptr = reinterpret_cast<const void*>(m_source.as_byte_span().data() + read_bytepos());
+               return {reinterpret_cast<const BlockT*>(ptr), blocks};
+            }
+
+            void advance(size_type bytes)
+               requires(is_byte_aligned())
+            {
+               m_read_bitpos += bytes * 8;
+               m_write_bitpos += bytes * 8;
+            }
+
+            template <std::unsigned_integral BlockT>
+               requires(is_byte_aligned())
+            size_t is_memory_aligned_to() const {
+               void* ptr = const_cast<void*>(static_cast<const void*>(m_source.as_byte_span().data() + read_bytepos()));
```

Comment:
Why the `const_cast` here? Probably worth a comment

Reply:
`std::align` is treating `ptr` as an out-param (taking it as `void*&` :fearful:). The data pointed to by `ptr` isn't changed, but it re-sets `ptr` to the next aligned address.